pilot/pkg/controller/workloadentry/workloadentry_controller.go

```diff
@@ -243,54 +244,76 @@ func (c *Controller) processNextWorkItem() bool {
 		return true
 	}
 
-	err := c.unregisterWorkload(workItem.entryName, workItem.proxy, workItem.disConTime, workItem.origConTime)
+	err := c.unregisterWorkload(workItem.entryName, workItem.proxy, workItem.disConTime, workItem.origConTime, workItem.autoCreated)
 	c.handleErr(err, item)
 	return true
 }
 
 func setConnectMeta(c *config.Config, controller string, conTime time.Time) {
+	if c.Annotations == nil {
+		c.Annotations = make(map[string]string)
+	}
 	c.Annotations[WorkloadControllerAnnotation] = controller
 	c.Annotations[ConnectedAtAnnotation] = conTime.Format(timeFormat)
 	delete(c.Annotations, DisconnectedAtAnnotation)
 }
 
 func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) error {
-	if !features.WorkloadEntryAutoRegistration || c == nil {
+	if c == nil {
 		return nil
 	}
-	// check if the WE already exists, update the status
-	entryName := autoregisteredWorkloadEntryName(proxy)
+	var entryName string
+	var autoCreate bool
+	if features.WorkloadEntryAutoRegistration &&
+		proxy.Metadata.AutoRegisterGroup != "" {
+		entryName = autoregisteredWorkloadEntryName(proxy)
+		autoCreate = true
+	} else if features.WorkloadEntryHealthChecks &&
+		proxy.Metadata.ProxyConfig != nil && proxy.Metadata.ProxyConfig.ReadinessProbe != nil {
+		entryName = lookupWorkloadEntryByProxyID(c.store, proxy)
```

Comment:
What if the wle has not been registered when proxy connect to pilot

Reply:
Or pilot has not yet discovered it

Reply:
The feature introduced by this PR - health checks for VMs that are not using auto-registration - implies that `WorkloadEntry` resource must exist beforehand.

However, I see your point that there is a room for an error due to something that a user has no control over.

I think in a scenario where auto-registration is not used, a user expects that a proxy must correspond to an explicitly created `WorkloadEntry`. If the `WorkloadEntry` is missing, it's a configuration mistake.

So, I would prefer to return an error out of `RegisterWorkload()` in case the `WorkloadEntry` is not found, which will cause Istio Proxy to re-connect:
* if it's a config propagation delay, the issue will eventually resolve transparently to the user
* if it's a human mistake, it will be visible to the user (as proxy will never come up)

Reply:
@hzxuzhonghu Instead of trying to infer the name of the corresponding `WorkloadEntry` from `proxy.ID`, I would prefer to introduce a new field into the node metadata specifically for that purpose.

E.g.,

```go
type NodeMetadata struct {
...
	// WorkloadEntry is the name of the `WorkloadEntry` this proxy corresponds to.
	WorkloadEntry string `json:"WORKLOAD_ENTRY,omitempty"`
...
}
```

This setting would be set for a sidecar through environment variable `ISTIO_META_WORKLOAD_ENTRY`.

WDYT?

Reply:
Sounds good, but this should be more generic, I remember we have a test for vm use some flag @stevenctl may know.
 grpc now use a different generator, not sure some other proxies in the future need this.

Reply:
added `WORKLOAD_ENTRY` to proxy metadata and return error if there is no such WorkloadEntry

@hzxuzhonghu PTAL

Reply:
Will take another look, sorry for the delay